sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/CoGroup.java

```diff
@@ -275,76 +358,263 @@ private FieldAccessDescriptor getFieldAccessDescriptor(TupleTag<?> tag) {
+
+    ExpandCrossProduct(JoinArguments joinArgs) {
+      this.joinArgs = joinArgs;
+    }
+
+    /**
+     * Select the following fields for the specified PCollection with the specified join args.
+     *
+     * <p>Each PCollection in the input must have fields specified for the join key.
+     */
+    public ExpandCrossProduct join(String tag, By clause) {
+      if (joinArgs.allInputsJoinArgs != null) {
+        throw new IllegalStateException("Cannot set both a global and per-tag fields.");
+      }
+      return new ExpandCrossProduct(joinArgs.with(tag, clause));
+    }
+
+    private Schema getOutputSchema(JoinInformation joinInformation) {
+      // Construct the output schema. It contains one field for each input PCollection, of type
+      // ROW. If a field supports outer-join semantics, then that field will be nullable in the
+      // schema.
+      Schema.Builder joinedSchemaBuilder = Schema.builder();
+      for (Map.Entry<String, Schema> entry : joinInformation.componentSchemas.entrySet()) {
+        FieldType fieldType = FieldType.row(entry.getValue());
+        if (joinArgs.getOuterJoinParticipation(entry.getKey())) {
+          fieldType = fieldType.withNullable(true);
+        }
+        joinedSchemaBuilder.addField(entry.getKey(), fieldType);
+      }
+      return joinedSchemaBuilder.build();
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollectionTuple input) {
+      JoinInformation joinInformation =
+          JoinInformation.from(input, joinArgs::getFieldAccessDescriptor);
+
+      Schema joinedSchema = getOutputSchema(joinInformation);
+
+      return joinInformation
+          .keyedPCollectionTuple
+          .apply("CoGroupByKey", CoGroupByKey.create())
+          .apply("Values", Values.create())
           .apply(
-              "extractKey" + tag,
+              "ExpandToRow",
               ParDo.of(
-                  new DoFn<T, KV<Row, Row>>() {
-                    @ProcessElement
-                    public void process(@Element Row row, OutputReceiver<KV<Row, Row>> o) {
-                      o.output(KV.of(Select.selectRow(row, keyFields, schema, keySchema), row));
-                    }
-                  }))
-          .setCoder(KvCoder.of(SchemaCoder.of(keySchema), SchemaCoder.of(schema)));
+                  new ExpandToRows(
+                      joinInformation.sortedTags,
+                      joinInformation.toRows,
+                      joinedSchema,
+                      joinInformation.tagToTupleTag)))
+          .setRowSchema(joinedSchema);
+    }
+
+    /** A DoFn that expands the result of a CoGroupByKey into the cross product. */
+    private class ExpandToRows extends DoFn<CoGbkResult, Row> {
+      private final List<String> sortedTags;
+      private final Map<String, SerializableFunction<Object, Row>> toRows;
+      private final Schema outputSchema;
+      private final Map<String, String> tagToTupleTag;
+
+      public ExpandToRows(
```

Comment:
This is an excellent point. Split this out so we create the iterators ahead of time. Also changed  maps to key off of tag index so that we don't have to hash the string on every single element.